### ISSUE TITLE
NavigatorKey is not mounted yet and the build context is null. 

### ISSUE THREAD

**Issue by mtgnoah** [Milad-Akarie/auto_route_library#437]
```class AuthGuard extends AutoRouteGuard {
  @override
  Future<bool> canNavigate(
      List<PageRouteInfo> routes, StackRouter router) async {
    // await Future.delayed(Duration(seconds: 2));
    final context = router.navigatorKey.currentContext;
    print(FirebaseAuth.instance.currentUser);
    // ignore: unnecessary_null_comparison
    if (Provider.of<User>(context, listen: false) == null) {
      print('first auth == null');
      await router.root.push(SignInPageRoute(onLoginResult: (loggedIn) {
        if (loggedIn) {
          print('first auth');
          router.pushAll(routes);
        } else {
          return SignInPage;
        }
      }));
      print('second auth == null');
      return false;
    }
    print('auth check from outside');
    return true;
  }
}
```
    if (Provider.of<User>(context, listen: false) == null) {

This line is causing the most issues with the error "The argument type 'BuildContext?' cannot be assigned to the parameter type 'BuildContext'.

**Comment by Milad-Akarie** [comment]
@mtgnoah  are you saying you're getting a null context? is this an initial page?

**Comment by mtgnoah** [comment]
@Milad-Akarie 
Yes I am getting a null context and this is an initial page.

**Comment by theweiweiway** [comment]
Hey @mtgnoah , this is inside your Route Guard right? I think this is the same issue as #290 - there are a few workarounds in that thread you can check out

**Comment by theweiweiway** [comment]
Closing due to duplicate issue

**Comment by mtgnoah** [comment]
@theweiweiway 
I ended up switching over to Riverpod and I think I can get it to work except for this context error so it won't start currently. Can I send you an invite for the repo?

**Comment by theweiweiway** [comment]
Sure

**Comment by mtgnoah** [comment]
@theweiweiway 
Sent an invite over. The branch is devroutes.

**Comment by theweiweiway** [comment]
Hey - I didn't get it for some reason - can you resend?

**Comment by mtgnoah** [comment]
Ok I resent it.

**Comment by theweiweiway** [comment]
@mtgnoah it looks alright - what's the error exactly? It doesn't look like there are any route guards so I"m not sure what context error you are talking about

**Comment by mtgnoah** [comment]
@theweiweiway 
Sorry about that I mislisted the issue there. It can't watch the auth service changes in the future part of main.dart. I was wondering if you could help me figure out how to fix that part or if it is a issue with the authentication_service.dart, my providers, or an idea on how to do it completely different. I'm open for any suggestions at this point because this issue has brought my development to a standstill.

**Comment by theweiweiway** [comment]
I think it has to do with how you are listening to auth state changes. I've never used riverpod and unfamiliar with it, so I can't help out there. But from what I know about firebase auth, I think u need to use the listener:


```dart
.authStateChanges().listen((User? user) {
    if (user == null) {
     setAuthStateAsSignedOut();
    } else {
     setAuthStateAsSignedIn();
    }
  });
```

Also, you will need to make sure you don't mount your `MaterialApp.router` until an authentication state has been assigned. Because initially when the app is starting the user won't be logged in or logged out. It'll be in an initial state.

What I do is in my state management, I have 3 states:
1. AuthState.initial
2. AuthState.signedIn
3. AuthState.signedOut

if `AuthState.initial`, render `SplashScreen`. Else, render `MaterialApp.router`.

**Comment by mtgnoah** [comment]
So now I changed a lot of stuff and your descriptions along with [https://github.com/rrousselGit/river_pod/issues/391](url) helped a ton but I get my splash screen now and can't figure out why. No errors except for one in the AuthenticationWrapper saying it might cause null to be returned. I have the AuthWrapper commented out because of this but even then I only get splash screen. I have also tried to pass just true or false to the authenticated in Myapp and that hasn't changed anything. Also if you know how I can return a bool with the isAuthenticated that might fix it.
